Issue 1929

garrettbaski opened the pull request:
This PR is probably too messy since I haven't contributed to this repo before and didn't have a must better way of adding this check with all of the Predicate/Token handling. But it should at least give the test cases and maybe some ideas on how to fix it to someone who knows the repo/language better.

See the issue here for some more info https://github.com/Azure/Azurite/issues/1929 but when querying table storage, if you have an entity with a not set field and you query on that field with a not equal, that entity will be returned since it is not equal. However this does not match the behavior in the real Azure Storage where if you query a specific field and that field is not set, it ignore that entity. 

For this PR, I basically added a check for any identifier to make sure the field is there doing 
``` item.properties.hasOwnProperty("<propertyValue">) ```
I'm not a fan of how it is jammed in. And there may be a better way to check the field, but I'm not super familiar with the code or typescript. 

The C# test tests the behavior. Current if ran against real Azure Storage, it will pass. However, if ran against Azurite it will fail. The added test in the table.entity.query.test.ts does the same thing

garrettbaski commented:
@microsoft-github-policy-service agree

notheotherben commented:
Hi @garrettbaski,

Looking at this change, it looks like there is a ton of repetitive code - that's usually a sign that there's a better way to tackle the problem (either through an abstraction, or through a common injection point). Walking through the parser logic here, it seems like there are a number of steps happening, and I'll try to capture my understanding here so that if someone more familiar with the code disagrees, they can ideally point out where I've led you astray :).

1. The parser consists of a state machine which consumes the query tokens and proceeds to convert these into one of `Identifier`, `Operator` or `Value` tokens (themselves contained in a `TokenMap`).
2. In the `StatePredicateFinished` parser state, these tuples are converted into an `IPredicate` (such as the `BooleanPredicate`) which provides the `convertPredicateForLokiJS` function. This function essentially performs the transform from a "query string token map" into a "JavaScript token map" (and you'll see this being done with the `r/this.push(Value|Identifier|Operator)/` methods in each predicate implementation).
3. The `convertPredicateForLokiJS` function is actually executed in the `StateQueryFinished` parser state, which is responsible for converting the whole shebang into a `transcribedQuery` string (which should be valid JavaScript that can then be used to construct a `Function` for execution).

Right now, you've approached solving this problem by doing two things:
- You've introduced a condition within each `IPredicate` implementation which attempts to add `item.properties.hasOwnProperty(...)` at either the start or end of the predicate's "JavaScript token map" when `convertPredicateForLokiJS` is called.
- Assuming that there can only be one `Identifier` token within a predicate (this is not necessarily true, as we can assert `x eq y`).

Given my understanding of the way this library works, I think you may be better served by focusing on the 3rd portion of the translation (the `StateQueryFinished` state)  and in particular the `onProcess` method there - where each predicate is converted into its JavaScript form.

I believe introducing the following changeset there might be all that is necessary to implement support for this, as well as to ensure that we don't run into problems with evaluating `x eq y` either.

```diff --git a/src/table/persistence/QueryTranscriber/StateQueryFinished.ts b/src/table/persistence/QueryTranscriber/StateQueryFinished.ts
index 227d0b4..74ac024 100644
--- a/src/table/persistence/QueryTranscriber/StateQueryFinished.ts
+++ b/src/table/persistence/QueryTranscriber/StateQueryFinished.ts
@@ -26,6 +26,13 @@ export default class StateQueryFinished implements IQPState {
     for (const taggedPredicate of context.taggedPredicates) {
       let predicate = "";
       if (taggedPredicate !== undefined) {
+        const requiredIdentifiers = taggedPredicate.tokenMap.tokens.filter(token => token.type.isIdentifier() && token.token !== "**blena**");
+
+        for (const identifierToken of requiredIdentifiers) {
+          predicate += " ";
+          predicate += `item.properties.hasOwnProperty("${identifierToken.token.replace(/\\/g, "\\\\")}") &&`;
+        }
+
         const convertedPredicate = taggedPredicate.convertPredicateForLokiJS();
         for (const taggedPredicateToken of convertedPredicate.tokenMap.tokens) {
           predicate += " ";
```

Doing so will result in any predicate which references an `Identifier` being prefixed with `item.properties.hasOwnProperty($$identifier$$) &&`, for each identifier used, and it avoids the need for you to introduce any additional changes to the `IPredicates`, greatly simplifying your changeset. As a last caution - making the above change will cause your unit tests to fail as the ordering that you've currently implemented (either pre-fixing or post-fixing the condition) no longer applies, I think this is likely for the best as there's no value in evaluating the predicate if the field does not exist.

garrettbaski commented:
> Hi @garrettbaski,
> 
> Looking at this change, it looks like there is a ton of repetitive code - that's usually a sign that there's a better way to tackle the problem (either through an abstraction, or through a common injection point). Walking through the parser logic here, it seems like there are a number of steps happening, and I'll try to capture my understanding here so that if someone more familiar with the code disagrees, they can ideally point out where I've led you astray :).
> 
> 1. The parser consists of a state machine which consumes the query tokens and proceeds to convert these into one of `Identifier`, `Operator` or `Value` tokens (themselves contained in a `TokenMap`).
> 2. In the `StatePredicateFinished` parser state, these tuples are converted into an `IPredicate` (such as the `BooleanPredicate`) which provides the `convertPredicateForLokiJS` function. This function essentially performs the transform from a "query string token map" into a "JavaScript token map" (and you'll see this being done with the `r/this.push(Value|Identifier|Operator)/` methods in each predicate implementation).
> 3. The `convertPredicateForLokiJS` function is actually executed in the `StateQueryFinished` parser state, which is responsible for converting the whole shebang into a `transcribedQuery` string (which should be valid JavaScript that can then be used to construct a `Function` for execution).
> 
> Right now, you've approached solving this problem by doing two things:
> 
> * You've introduced a condition within each `IPredicate` implementation which attempts to add `item.properties.hasOwnProperty(...)` at either the start or end of the predicate's "JavaScript token map" when `convertPredicateForLokiJS` is called.
> * Assuming that there can only be one `Identifier` token within a predicate (this is not necessarily true, as we can assert `x eq y`).
> 
> Given my understanding of the way this library works, I think you may be better served by focusing on the 3rd portion of the translation (the `StateQueryFinished` state) and in particular the `onProcess` method there - where each predicate is converted into its JavaScript form.
> 
> I believe introducing the following changeset there might be all that is necessary to implement support for this, as well as to ensure that we don't run into problems with evaluating `x eq y` either.
> 
> ```diff
> index 227d0b4..74ac024 100644
> --- a/src/table/persistence/QueryTranscriber/StateQueryFinished.ts
> +++ b/src/table/persistence/QueryTranscriber/StateQueryFinished.ts
> @@ -26,6 +26,13 @@ export default class StateQueryFinished implements IQPState {
>      for (const taggedPredicate of context.taggedPredicates) {
>        let predicate = "";
>        if (taggedPredicate !== undefined) {
> +        const requiredIdentifiers = taggedPredicate.tokenMap.tokens.filter(token => token.type.isIdentifier() && token.token !== "**blena**");
> +
> +        for (const identifierToken of requiredIdentifiers) {
> +          predicate += " ";
> +          predicate += `item.properties.hasOwnProperty("${identifierToken.token.replace(/\\/g, "\\\\")}") &&`;
> +        }
> +
>          const convertedPredicate = taggedPredicate.convertPredicateForLokiJS();
>          for (const taggedPredicateToken of convertedPredicate.tokenMap.tokens) {
>            predicate += " ";
> ```
> 
> Doing so will result in any predicate which references an `Identifier` being prefixed with `item.properties.hasOwnProperty($$identifier$$) &&`, for each identifier used, and it avoids the need for you to introduce any additional changes to the `IPredicates`, greatly simplifying your changeset. As a last caution - making the above change will cause your unit tests to fail as the ordering that you've currently implemented (either pre-fixing or post-fixing the condition) no longer applies, I think this is likely for the best as there's no value in evaluating the predicate if the field does not exist.

Yeah this is a much better way of doing it. I couldn't think of how to do it in a single spot. I changed it to this and fixed up all the tests.

notheotherben commented:
Looks solid to me, thanks Garrett - hopefully this is a bit quicker for @edwin-huber to review now, shout if there's anything else I can help with (unfortunately being an approver is not one of those things right now).

blueww commented:
@edwin-huber , @XiaoningLiu 
Would you please help to review if the fix is good to you?

garrettbaski commented:
I think this will be invalidated by https://github.com/Azure/Azurite/pull/1975 if that is merged first. In that case, I just need to ensure that the test case that I have here successfully executes and if so then this will not be needed.

blueww commented:
@garrettbaski 

After https://github.com/Azure/Azurite/pull/1975 merged, if the issue of this PR still exist?
If so, would you please update this PR based on the current code base?

Thanks for your contribution!

blueww reviewed (approved):
Only test case change. Looks good to me.

blueww commented:
@XiaoningLiu  , @EmmaZhu 
Any comments for this PR?
